From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Chad Brubaker <cbrubaker@google.com>
Date: Wed, 16 Dec 2015 14:55:53 -0800
Subject: [PATCH 15/32] DO NOT MERGE Cache intermediate CA separately

Intermediate CAs are cached in order to support servers that fail to
sent a complete chain to a root. These certificates should be cached to
support these servers but these certificates must not be trusted as
trust anchors. Store them separately to prevent confusion between
trusted roots and cached intermediates.

(cherry-picked from commit 198aca1fb638a2a98e89fb9f284108ad576d0c3b)
Bug: 26232830
Change-Id: I520f50729b55fc7412c7d133335bc9e3c190bbf6
(cherry picked from commit 1cca3cd728e05c18ae3376cf39a8503b736d87b1)
(cherry picked from commit 02e4d2251489d2393b0394e7a38ec093e9bc8243)

Android 4.2 port (android-4.2.2_r1) of patch 0015 of the 4.4 series,
made from its Android 4.3 port. Paths and names: libcore/crypto of 4.4
(org.conscrypt, org_conscrypt_NativeCrypto.cpp in libjavacrypto) is
libcore/luni on 4.2 (org.apache.harmony.xnet.provider.jsse,
org_apache_harmony_xnet_provider_jsse_NativeCrypto.cpp in libjavacore).
4.2 lacks libcore 608ba9e (Android 4.3, "Do not include bogus certs in
final chain output"): the anchor loop of
cleanupCertChainAndFindTrustAnchors() runs to chain.length, not to
currIndex. With this change the chain array can grow (Arrays.copyOf
doubles it and leaves nulls at the end), and that loop would pass null
to findTrustAnchorBySubjectAndPublicKey(). This patch therefore also
takes the one-line bound of 608ba9e (anchorIndex <= currIndex, as on 4.3
and 4.4) and its test, TrustManagerImplTest.testDuplicateInChain.
Otherwise the same change.
---
 .../xnet/provider/jsse/TrustManagerImpl.java  | 41 +++++++++++++++----
 .../provider/jsse/TrustManagerImplTest.java   | 16 ++++++++
 2 files changed, 50 insertions(+), 7 deletions(-)

diff --git a/luni/src/main/java/org/apache/harmony/xnet/provider/jsse/TrustManagerImpl.java b/luni/src/main/java/org/apache/harmony/xnet/provider/jsse/TrustManagerImpl.java
index 021824994f8..3a8db62a7ef 100644
--- a/luni/src/main/java/org/apache/harmony/xnet/provider/jsse/TrustManagerImpl.java
+++ b/luni/src/main/java/org/apache/harmony/xnet/provider/jsse/TrustManagerImpl.java
@@ -67,11 +67,16 @@ public final class TrustManagerImpl implements X509TrustManager {
     private final CertPathValidator validator;
 
     /**
-     * An index of TrustAnchor instances that we've seen. Unlike the
-     * TrustedCertificateStore, this may contain intermediate CAs.
+     * An index of TrustAnchor instances that we've seen.
      */
     private final TrustedCertificateIndex trustedCertificateIndex;
 
+    /**
+     * An index of intermediate certificates that we've seen. These certificates are NOT implicitly
+     * trusted and must still form a valid chain to an anchor.
+     */
+    private final TrustedCertificateIndex intermediateIndex;
+
     /**
      * This is lazily initialized in the AndroidCAStore case since it
      * forces us to bring all the CAs into memory. In the
@@ -140,6 +145,7 @@ public final class TrustManagerImpl implements X509TrustManager {
         this.validator = validatorLocal;
         this.factory = factoryLocal;
         this.trustedCertificateIndex = trustedCertificateIndexLocal;
+        this.intermediateIndex = new TrustedCertificateIndex();
         this.acceptedIssuers = acceptedIssuersLocal;
         this.err = errLocal;
     }
@@ -288,7 +294,7 @@ public final class TrustManagerImpl implements X509TrustManager {
             // will have been removed in
             // cleanupCertChainAndFindTrustAnchors.  http://b/3404902
             for (int i = 1; i < newChain.length; i++) {
-                trustedCertificateIndex.index(newChain[i]);
+                intermediateIndex.index(newChain[i]);
             }
         } catch (InvalidAlgorithmParameterException e) {
             throw new CertificateException(e);
@@ -349,9 +355,30 @@ public final class TrustManagerImpl implements X509TrustManager {
             }
         }
 
-        // 2. Find the trust anchor in the chain, if any
+        // 2. Add any missing intermediates to the chain
+        while (true) {
+            TrustAnchor nextIntermediate =
+                    intermediateIndex.findByIssuerAndSignature(chain[currIndex]);
+            if (nextIntermediate == null) {
+                break;
+            }
+            // Append intermediate
+            X509Certificate cert = nextIntermediate.getTrustedCert();
+            // don't mutate original chain, which may be directly from an SSLSession
+            if (chain == original) {
+                chain = original.clone();
+            }
+            // Grow the chain if needed
+            if (currIndex == chain.length - 1) {
+                chain = Arrays.copyOf(chain, chain.length * 2);
+            }
+            chain[currIndex + 1] = cert;
+            currIndex++;
+        }
+
+        // 3. Find the trust anchor in the chain, if any
         int anchorIndex;
-        for (anchorIndex = 0; anchorIndex < chain.length; anchorIndex++) {
+        for (anchorIndex = 0; anchorIndex <= currIndex; anchorIndex++) {
             // If the current cert is a TrustAnchor, we can ignore the rest of the chain.
             // This avoids including "bridge" CA certs that added for legacy compatibility.
             TrustAnchor trustAnchor = findTrustAnchorBySubjectAndPublicKey(chain[anchorIndex]);
@@ -361,13 +388,13 @@ public final class TrustManagerImpl implements X509TrustManager {
             }
         }
 
-        // 3. If the chain is now shorter, copy to an appropriately sized array.
+        // 4. If the chain is now shorter, copy to an appropriately sized array.
         int chainLength = anchorIndex;
         X509Certificate[] newChain = ((chainLength == chain.length)
                                       ? chain
                                       : Arrays.copyOf(chain, chainLength));
 
-        // 4. If we didn't find a trust anchor earlier, look for one now
+        // 5. If we didn't find a trust anchor earlier, look for one now
         if (trustAnchors.isEmpty()) {
             TrustAnchor trustAnchor = findTrustAnchorByIssuerAndSignature(newChain[anchorIndex-1]);
             if (trustAnchor != null) {
diff --git a/luni/src/test/java/org/apache/harmony/xnet/provider/jsse/TrustManagerImplTest.java b/luni/src/test/java/org/apache/harmony/xnet/provider/jsse/TrustManagerImplTest.java
index fe5f4f08136..9757bc5a358 100644
--- a/luni/src/test/java/org/apache/harmony/xnet/provider/jsse/TrustManagerImplTest.java
+++ b/luni/src/test/java/org/apache/harmony/xnet/provider/jsse/TrustManagerImplTest.java
@@ -98,6 +98,22 @@ public class TrustManagerImplTest extends TestCase {
         assertValid(chain1, tm);
     }
 
+    // We should ignore duplicate cruft in the certificate chain
+    // See https://code.google.com/p/android/issues/detail?id=52295 http://b/8313312
+    public void testDuplicateInChain() throws Exception {
+        // chain3 should be server/intermediate/root
+        KeyStore.PrivateKeyEntry pke = TestKeyStore.getServer().getPrivateKey("RSA", "RSA");
+        X509Certificate[] chain3 = (X509Certificate[])pke.getCertificateChain();
+        X509Certificate root = chain3[2];
+        X509Certificate intermediate = chain3[1];
+        X509Certificate server = chain3[0];
+
+        X509Certificate[] chain4 = new X509Certificate[] { server, intermediate,
+                                                           server, intermediate
+        };
+        assertValid(chain4, trustManager(root));
+    }
+
     public void testGetFullChain() throws Exception {
         // build the trust manager
         KeyStore.PrivateKeyEntry pke = TestKeyStore.getServer().getPrivateKey("RSA", "RSA");
